Objects/unicodeobject.c

```diff
@@ -14855,6 +14882,9 @@ PyUnicode_InternFromString(const char *cp)
 void
 _PyUnicode_ClearInterned(PyInterpreterState *interp)
 {
+    if (has_shared_intern_dict(interp)) {
+        return; // the dict doesn't belong to this interpreter
+    }
```

Comment:
As with `clear_interned_dict()`, we also need to set the `PyInterpreterState` field to NULL:

```suggestion
    if (has_shared_intern_dict(interp)) {
        // The dict doesn't belong to this interpreter
        // so we don't otherwise clean it up.
        _Py_INTERP_CACHED_OBJECT(interp, interned_strings) = NULL;
        return;
    }
```

Ideally we could call `clear_interned_dict()` directly:

```suggestion
    if (has_shared_intern_dict(interp)) {
        clear_interned_dict();
        return;
    }
```

or maybe put a `goto` down to the call at the bottom of this function.

I suppose the whole function could look like this:

```C
void
_PyUnicode_ClearInterned(PyInterpreterState *interp)
{
    PyObject *interned = get_interned_dict(interp);
    if (interned == NULL) {
        return;
    }
    assert(PyDict_CheckExact(interned));

#ifdef Py_DEBUG
    if (!has_shared_intern_dict(interp)) {
        ...
    }
#endif
    clear_interned_dict(interp);
}
```

However, that's maybe too much churn for targeting 3.12.  Perhaps in the main branch...